era_5g_interface/task_handler_internal_q.py

```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
                 The format is NetApp-specific.
             image (_type_): The image to be processed.
         """
+
+        # make sure multiple calls of store_image do not interfere with each other
+        self._q_put_lock.acquire()
+
         self.frame_id += 1
         # logger.info(f"TaskHandlerInternalQ received frame id: {self.frame_id} with timestamp: {metadata['timestamp']}")
         try:
             self._q.put((metadata, image), block=False)
         except Full:
-            pass
-            # TODO: raise an exception
+            if self.if_queue_full == "discard_new":
+                pass  # do nothing (discard the new data)
+            elif self.if_queue_full == "discard_oldest":
+                # remove the oldest data from queue to make up space for the new one,
+                # but make sure not to discard control commands
+                self._q.mutex.acquire()  # acquire the queue lock to prevent it from mutating during the removal
+                try:
+                    space_available = False
+                    type_of_old_data = type(self._q.queue[0])
+                    if type_of_old_data is not ControlCommand:
+                        self._q._get()  # remove the old data from queue
+                        space_available = True
+                    else:
+                        logger.warning(
+                            "Discarding new data instead of the old one to prevent losing a control command in queue."
+                        )
+                        # this situation could also raise an exception, if desired
+                finally:
+                    self._q.mutex.release()
+                if space_available:
+                    self._q.put((metadata, image), block=False)
+            else:
+                raise RuntimeError("Internal queue is full and no data was allowed to be discarded.")
```

Comment:
Why `RuntimeError`? I would rather use some custom exception.

Reply:
Changed to QueueDiscardError